cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py

```diff
@@ -534,8 +534,14 @@ def set_final_status():
         hookenv.status_set('waiting', 'Waiting to retry addon deployment')
         return
 
-    if addons_configured and not all_kube_system_pods_running():
-        hookenv.status_set('waiting', 'Waiting for kube-system pods to start')
+    unready = get_kube_system_pods_not_running()
+    if addons_configured and (unready is None or len(unready) > 0):
+        if unready is None:
+            msg = 'Waiting for kube-system pods to start'
+        elif len(unready) > 0:
+            msg = 'Waiting for {} kube-system pod{} to start'
+            msg = msg.format(len(unready), "s"[len(unready) == 1:])
```

Comment:
The readability of this code really took a hit. I'm not a fan.

What does `None` represent? I have to go read `get_kube_system_pods_not_running` to find out. This is why I don't like sentinel values - they don't tell you what they represent. Exceptions do, and IMO that would improve readability quite a bit:

```
try:
    unready = get_kube_system_pods_not_running()
except FailedToGetKubeSystemPodStatus:
    msg = "Waiting for kube-system pods to start"
    hookenv.status_set('waiting', msg)
    return

if unready:
    msg = 'Waiting for {} kube-system pod{} to start'
    msg = msg.format(len(unready), "s"[len(unready) == 1:])
    hookenv.status_set('waiting', msg)
    return
```

But even then, is this change worth it? What's the motivation behind this?

Reply:
I struggled with the return value options and discussed using an exception vs returning a completely different type. I don't like the idea of not knowing what a function is going to return. The exception does make it far more obvious what is going on here and I like that. If we continue with this part of the change, I would want to change it as you suggest here.

The goal was to provide more information, the count of unready pods, than just "waiting for system pods". This isn't an important change, but I felt value to knowing if we are making progress or if we're still waiting on the same 5 pods as last time. You can get this information from the debug-log, but it seems a simple thing that requires very little knowledge of the user. It was just something I thought about while I was debugging pods not coming up as being useful to surface.

Reply:
> ... I felt value to knowing if we are making progress or if we're still waiting on the same 5 pods as last time.

I can get behind that. +1

Reply:
The retry decorator doesn't look like it will retry on an exception. This may influence our design on the return value/exceptions depending on the decision wrt the retry decorator.